### ISSUE TITLE
should run and test default to listing instead of running all cases and have a --all switch?

### ISSUE THREAD

**Issue by urbanjost** [fortran-lang/fpm#245]
I find a more intuitive behavior for `run` and `test` if that if there is just one target it runs it; but if there is more than one it just lists the candidates and that you either have to use a new switch --all to get the current behavior or give a list of targets. Does that seem like a necessary change to anyone else? Adding the switch is not much of a burden and if you have dozens of apps or tests right now the only way to get a nice list is to enter a name that does not exist. Another option might be to have a switch like -toc that lists the candidates in the nice table you get when you enter a bad name.

**Comment by awvwgk** [comment]
The other way round it makes more sense to me. I would use a ``--list`` option to get all possible targets. I think at least for test that should show the integrity of the project I would prefer to run all tests rather than getting a list (and exit status zero).

Consider a CI setup with fpm, both `fpm test` and `fpm test --all` would return a zero exit status, but the former would only work as expected for projects with a single test target. Together with the automatic discovery of test executables this might lead to surprising outcomes.

But it can certainly dependent on the design of the project. After I started using fpm more seriously in my projects, I reconsidered some of my project designs for fpm, leading to overall smaller projects with usually a single executable target per project.

**Comment by LKedward** [comment]
I agree that `fpm test` should run all tests by default, but I agree with @urbanjost with respect to `fpm run`; I really don't like having `fpm run` behave the same way as `fpm test`. Some libraries come with *many* example programs and it doesn't make sense to invoke them all sequentially. I also find myself purposely invoking `fpm run invalid_name` to get a nice list of programs so that I can run them one by one.

What does everyone think to changing `fpm run` to require a program name argument if there is more than one program/example available and using flags to run all programs and/or examples sequentially?

```sh
$ fpm run NAME | [--programs] [--examples] 
```

where `NAME` can be the executable name for any program or example.

I think `cargo run` has a similar requirement when more than one executable target is available (but their selection syntax is too verbose IMO).

**Comment by urbanjost** [comment]
Yes Thumbs up.

## CONSIDER A MENU MODE
I have another alternative I have experimented with -- menus.

 I have an alpha-alpha version of my own that:
1) has the Fortran intrinsics available via `help`
2) looks for a command in the search path called fpm-NAME when an unknown subcommand is called
3) allows for "response files" as described in the M_CLI2 module
4) allows for a __-i__ switch on `run` and `test`
5) starts a shell-like interface if no options are specified on `fpm` (very basic)
none of the ideas I played with my private "alpha-alpha" version were show-stoppers that I thought had a priority above the bigger things that were going on like auto-dependencies (and I think are all still behind allowing compiler and load options and a command-line  way of adding packages to the search list and having something like `avpkg` built-in  (I renamed my copy of avpkg `fpm-search` and call it with 2) above, by the way) but when it looked like this proposal was not going anywhere I  become fond of "4)" -- the -i switch is for "interactive" mode (maybe it should be -m for "menu" mode) but it produces a menu list of the executables and you can pick them numerically. Just mentioning it as an alternative idea; my private version diverged enough it would probably be easier to reintroduce it in the main code than to pull it from it.

## STILL OCCASIONALLY WANT TO RUN ALL
 I did not pick up on your idea of just making `test` and `run` behave differently, which seems such an obvious solution after the fact!  So all for it. I agree that I want to run all the tests the majority of the time; usually I just want to run one only when I am writting one.
 I do occasionally want to run everything  ( one example is a graphics library where I want to pick a file output format and run all the examples, which there are dozens of).  

So like you said except add 
```bash
fpm run '*'
fpm run '*' --example
```
as special cases where  you want to run all the cases. 
## FANCIER MATCHING?
Maybe globbing or RE or even a --all switch would be better?

Since '*' is really a glob expression note there is  a simple globbing routine in M_strings  called matchw(just does * and ? not [...] and such) that would be a nice compromise and that I think most users would already be familiar with the syntax of. There is also a routine of the same name by Clive Page available in a lot of places including the Fortran Wiki but I am not sure of the licensing.

Or would it be nice for that (outlier?) case if the names could be Regular Expressions since you have to quote the * anyway?
I started a Fortran-based BRE module called M_match based on the original Ratfor version for those times where depending on the C libraries was not desirable but it needs work (PS: pretty sure Regular Expressions were first written in that Fortran variant -- ironic it is so hard to find a Fortran version now-adays).  It would work well enough to be able to say things like 'fpm run '.*'` for now though, even as if as a place-holder for something better someday.

 In lieu of that I would still like a --all switch for those admittedly rare cases where I want to run all the examples (sometimes just to make sure they all still run, more like a test than an example). 

I think the '*' is nice and avoids bloating the `run` command with rarely used options. But since it needs quoted maybe going the whole way to RE would add greater flexibility but might be overkill unless there are other options where it would be useful, like searching for packages or through the help.  The -all switch needs no quoting.  

There is also the case of having some alphameric reserved name like ALL acting as a flag, but I rarely like that method, and I think has drawbacks here.

## SUMMARY
So good with just changing `run` as you described; throwing out the "menu" idea to see if there is any interest", and would like a way to occasionally run them all without having to name them all (I have some packages with dozens of example names I would hate to have to name individually when using them for a "confidence test") . Probably just allowing '*' to mean all would be the simplest  -- and also keep `run` from getting bloated with rarely-used options; We could get fancier using globbing or RE but not sure if anyone needs that.

**Comment by LKedward** [comment]
Some great ideas @urbanjost. Yes an interactive menu code would definitely be useful when you want to select from many programs; I think it would be a good candidate for the proposed plugins mechanism once the *fpm* model is accessible.

I hadn't thought of RE matching but it is something that would certainly be useful to run subsets of tests/programs. And '*' is nice and short compared to a flag.

**Comment by urbanjost** [comment]
## CARGO
Looked up cargo(1) and it does something similar as mentioned, where if there is only one target it is used unspecified, otherwise it has to be listed. Not sure if it lists available targets or can take RE or globbing for names. Curious about that and if there if anyone is using anything else with similiar
requirements and how it is handled. 

```text
If neither `--bin` nor `--example` are given, then if the package only has one
bin target it will be run. Otherwise `--bin` specifies the bin target to run,
and `--example` specifies the example target to run. At most one of `--bin` or
`--example` can be provided.
```
